hugr-core/src/extension/prelude/array.rs

```diff
@@ -312,6 +331,192 @@ pub fn new_array_op(element_ty: Type, size: u64) -> ExtensionOp {
+    /// To avoid recursion when defining the extension, take the type definition as an argument.
+    fn signature_from_def(&self, array_def: &TypeDef) -> SignatureFunc {
+        // array<N, T1>, (T1, *A -> T2, *A), -> array<N, T2>, *A
+        let params = vec![
+            TypeParam::max_nat(),
+            TypeBound::Any.into(),
+            TypeBound::Any.into(),
+            TypeParam::new_list(TypeBound::Any),
+        ];
+        let n = TypeArg::new_var_use(0, TypeParam::max_nat());
+        let t1 = Type::new_var_use(1, TypeBound::Any);
+        let t2 = Type::new_var_use(2, TypeBound::Any);
+        let s = TypeRV::new_row_var_use(3, TypeBound::Any);
+        PolyFuncTypeRV::new(
+            params,
+            FuncTypeBase::<RowVariable>::new(
+                vec![
+                    instantiate(array_def, n.clone(), t1.clone()).into(),
+                    Type::new_function(FuncTypeBase::<RowVariable>::new(
+                        vec![t1.into(), s.clone()],
+                        vec![t2.clone().into(), s.clone()],
+                    ))
+                    .into(),
+                    s.clone(),
+                ],
+                vec![instantiate(array_def, n, t2).into(), s],
+            ),
+        )
+        .into()
+    }
+}
+
+impl MakeOpDef for ArrayScanDef {
+    fn from_def(op_def: &OpDef) -> Result<Self, OpLoadError>
+    where
+        Self: Sized,
+    {
+        crate::extension::simple_op::try_from_name(op_def.name(), op_def.extension())
+    }
+
+    fn signature(&self) -> SignatureFunc {
+        self.signature_from_def(array_type_def())
+    }
+
+    fn extension(&self) -> ExtensionId {
+        PRELUDE_ID
+    }
+
+    fn description(&self) -> String {
+        "A combination of map and foldl. Applies a function to each element \
+        of the array with an accumulator that is passed trough from left to \
+        right. Returns the resulting array and the final state of the \
+        accumulator."
+            .into()
+    }
+
+    /// Add an operation implemented as an [MakeOpDef], which can provide the data
+    /// required to define an [OpDef], to an extension.
+    //
+    // This method is re-defined here since we need to pass the array type def while
+    //  computing the signature, to avoid recursive loops initializing the extension.
+    fn add_to_extension(
+        &self,
+        extension: &mut Extension,
+    ) -> Result<(), crate::extension::ExtensionBuildError> {
+        let sig = self.signature_from_def(extension.get_type(ARRAY_TYPE_NAME).unwrap());
+        let def = extension.add_op(self.name(), self.description(), sig)?;
+
+        self.post_opdef(def);
+
+        Ok(())
+    }
+}
+
+/// Definition of the array scan op.
+#[derive(Clone, Debug, PartialEq)]
+pub struct ArrayScan {
+    /// The element type of the input array.
+    src_ty: Type,
+    /// The target element type of the output array.
+    tgt_ty: Type,
+    /// The accumulator types.
+    acc_tys: Vec<Type>,
```

Comment:
It's not clear to me why this is an array of types and not just a type -- isn't the accumulator a single type?

Reply:
I used a row variable for the accumulator. Setting the accumulator to the empty row recovers the signature of `map`